src/indexes/prefix.rs

```diff
@@ -0,0 +1,636 @@
+        let calculated_len = 1 + sub_names.len();
+        let mut combined: Vec<&[u8]> = Vec::with_capacity(calculated_len);
+        combined.push(top_name);
+        combined.extend(sub_names.iter().map(|sub_name| sub_name.as_ref()));
+        debug_assert_eq!(calculated_len, combined.len()); // as long as we calculate correctly, we don't need to reallocate
+        let storage_prefix = to_length_prefixed_nested(&combined);
+        IndexPrefix {
+            storage_prefix,
+            data: PhantomData,
+            pk_name: pk_name.to_vec(),
+            de_fn_kv,
+            de_fn_v,
+        }
+    }
+}
+
+impl<'b, K, T, B> IndexPrefix<K, T, B>
+where
+    B: PrimaryKey<'b>,
+    K: KeyDeserialize,
+    T: Serialize + DeserializeOwned,
+{
+    pub fn range_raw<'a>(
+        &self,
+        store: &'a dyn Storage,
+        min: Option<Bound<'b, B>>,
+        max: Option<Bound<'b, B>>,
+        order: Order,
+    ) -> Box<dyn Iterator<Item = StdResult<Record<T>>> + 'a>
+    where
+        T: 'a,
+    {
+        let de_fn = self.de_fn_v;
+        let pk_name = self.pk_name.clone();
+        let mapped = range_with_prefix(
+            store,
+            &self.storage_prefix,
+            min.map(|b| b.to_raw_bound()),
+            max.map(|b| b.to_raw_bound()),
+            order,
+        )
+        .map(move |kv| (de_fn)(store, &pk_name, kv));
+        Box::new(mapped)
+    }
+
+    pub fn keys_raw<'a>(
+        &self,
+        store: &'a dyn Storage,
+        min: Option<Bound<'b, B>>,
+        max: Option<Bound<'b, B>>,
+        order: Order,
+    ) -> Box<dyn Iterator<Item = Vec<u8>> + 'a> {
+        keys_with_prefix(
+            store,
+            &self.storage_prefix,
+            min.map(|b| b.to_raw_bound()),
+            max.map(|b| b.to_raw_bound()),
+            order,
+        )
+    }
+
+    /// Clears the prefix, removing the first `limit` elements (or all if `limit == None`).
+    pub fn clear(&self, store: &mut dyn Storage, limit: Option<usize>) {
+        const TAKE: usize = 10;
+        let mut cleared = false;
+
+        let mut left_to_clear = limit.unwrap_or(usize::MAX);
+
+        while !cleared {
+            // Take just TAKE elements to prevent possible heap overflow if the prefix is big,
+            // but don't take more than we want to clear.
+            let take = TAKE.min(left_to_clear);
+
+            let paths = keys_full(store, &self.storage_prefix, None, None, Order::Ascending)
+                .take(take)
+                .collect::<Vec<_>>();
+
+            for path in &paths {
+                store.remove(path);
+            }
+            left_to_clear -= paths.len();
+
+            cleared = paths.len() < take || left_to_clear == 0;
+        }
+    }
+
+    /// Returns `true` if the prefix is empty.
+    pub fn is_empty(&self, store: &dyn Storage) -> bool {
+        range_full(store, &self.storage_prefix, None, None, Order::Ascending)
+            .next()
+            .is_none()
```

Comment:
This could also be optimized to use `storage.range_keys`, right? Looks like I missed that in the first PR.